
 pub const ARGUMENT_REGISTERS: [X86Operand; 6] = [
@@ -110,11 +111,12 @@ fn map_operand(
                 .expect("expected array entry")
                 .typ
                 .clone();
+
             // println!("Matching operand for {:?}", index);
             let idx_op = map_operand(method_cfg, index, x86_instructions, globals);
             // println!("index operand is  {:?}", idx_op);
 
-            let index_reg: Register = reg_for_type(Register::R10, &Type::Int);
+            let index_reg: Register = reg_for_type(Register::Rax, &Type::Int);
 
             // Load base address of array into R11
             x86_instructions.push(X86Insn::Lea(
@@ -146,7 +148,7 @@ fn map_operand(
             X86Operand::Address(
                 None,
                 Some(Register::R11),
-                Register::R10,
+                Register::Rax,
                 element_size,
                 array_typ,
             )
@@ -165,7 +167,7 @@ fn map_operand(
                 .clone();
             let idx_op = map_operand(method_cfg, index, x86_instructions, globals);
 
-            let index_reg: Register = reg_for_type(Register::R10, &Type::Int);
+            let index_reg: Register = reg_for_type(Register::Rax, &Type::Int);
 
             x86_instructions.push(X86Insn::Mov(
                 idx_op,
@@ -188,7 +190,7 @@ fn map_operand(
             X86Operand::Address(
                 Some(name.to_string()),
                 None,
-                Register::R10,
+                Register::Rax,
                 element_size,
                 typ.clone(),
             )
@@ -272,7 +274,7 @@ fn add_instruction(
                     // support loading a 64-bit constant directly
                     x86_instructions.push(X86Insn::Mov(
                         X86Operand::Constant(((src.clone() as u64) & 0xFFFFFFFF) as i64),
-                        X86Operand::Reg(Register::R10),
+                        X86Operand::Reg(Register::R11),
                         Type::Long,
                     )); // lower 32 bits
                     x86_instructions.push(X86Insn::Mov(
@@ -286,11 +288,11 @@ fn add_instruction(
                     ));
                     x86_instructions.push(X86Insn::Or(
                         X86Operand::Reg(Register::Rax),
-                        X86Operand::Reg(Register::R10),
+                        X86Operand::Reg(Register::R11),
                     ));
                     let dest_location = map_operand(method_cfg, dest, x86_instructions, globals);
                     x86_instructions.push(X86Insn::Mov(
-                        X86Operand::Reg(Register::R10),
+                        X86Operand::Reg(Register::R11),
                         dest_location,
                         Type::Long,
                     ));
diff --git a/src/regalloc.rs b/src/regalloc.rs
index be8aa7f..94056e5 100644
--- a/src/regalloc.rs
+++ b/src/regalloc.rs
@@ -722,7 +722,7 @@ pub fn reg_alloc(method_cfgs: &mut BTreeMap<String, CFG>, globals: &BTreeMap<Str
         X86Operand::Reg(Register::R8),   // 5th argument
         X86Operand::Reg(Register::R9),   // 6th argument
 
-        // X86Operand::Reg(Register::R10),  // Scratch (caller-saved temp), rarely reserved by ABI
+        X86Operand::Reg(Register::R10),  // Scratch (caller-saved temp), rarely reserved by ABI
         // X86Operand::Reg(Register::R11),  // Scratch (caller-saved temp), rarely reserved by ABI
 
         // // Callee saved (non-volatile) — must be preserved by callee across calls
diff --git a/test b/test
index ba18daf..0bcd650 100755
Binary files a/test and b/test differ
diff --git a/test.s b/test.s
index e4eff81..4c950ac 100644
--- a/test.s
+++ b/test.s
@@ -1,17