diff --git a/baz.dot b/baz.dot
index fcee39f..0e06c52 100644
--- a/baz.dot
+++ b/baz.dot
@@ -1,4 +1,4 @@
 digraph CFG {
   node [shape=box fontname="monospace"];
-  0 [label="Block 0\l_t0 <- 1\le <- e + _t0\l_t1 <- String(1, String)\lprintf(_t1, e)\lret e\l"];
+  0 [label="Block 0\l_t0 <- 1\le <- e + _t0\l_t1 <- String(0, String)\lprintf(_t1, e)\lret e\l"];
 }
diff --git a/baz.png b/baz.png
index c1aea4c..f16e1d9 100644
Binary files a/baz.png and b/baz.png differ
diff --git a/cond.dot b/cond.dot
index 2cd943f..de7236e 100644
--- a/cond.dot
+++ b/cond.dot
@@ -1,4 +1,4 @@
 digraph CFG {
   node [shape=box fontname="monospace"];
-  0 [label="Block 0\l_t0 <- String(0, String)\l_t1 <- 1\lprintf(_t0, _t1)\lret 1\l"];
+  0 [label="Block 0\l_t0 <- String(1, String)\l_t1 <- 1\lprintf(_t0, _t1)\lret 1\l"];
 }
diff --git a/cond.png b/cond.png
index 402b0e8..4f36f91 100644
Binary files a/cond.png and b/cond.png differ
diff --git a/do.dot b/do.dot
index 48d55a5..f403616 100644
--- a/do.dot
+++ b/do.dot
@@ -7,10 +7,10 @@ digraph CFG {
   4 [label="Block 4\l_t6 <- String(0, String)\lprintf(_t6)\lujmp do6\l"];
   5 [label="Block 5\l_t7 <- String(1, String)\lprintf(_t7)\lujmp do6\l"];
   6 [label="Block 6\lujmp do3\l"];
-  0 -> 2 [label="False"];
   0 -> 1 [label="True"];
-  1 -> 4 [label="True"];
+  0 -> 2 [label="False"];
   1 -> 5 [label="False"];
+  1 -> 4 [label="True"];
   2 -> 3 [label=""];
   4 -> 6 [label=""];
   5 -> 6 [label=""];
diff --git a/factorial.dot b/factorial.dot
index 3ed2563..aea3e2c 100644
--- a/factorial.dot
+++ b/factorial.dot
@@ -5,6 +5,6 @@ digraph CFG {
   2 [label="Block 2\l_t4 <- 1\l_t5 <- _t0 - _t4\l_t6 <- factorial(_t5)\l_t7 <- _t0 * _t6\lret _t7\l"];
   3 [label="Block 3\lexit(-1)\l"];
   -1 -> 3 [label=""];
-  0 -> 2 [label="False"];
   0 -> 1 [label="True"];
+  0 -> 2 [label="False"];
 }
diff --git a/fib.dot b/fib.dot
index ab08e4b..4193d67 100644
--- a/fib.dot
+++ b/fib.dot
@@ -4,6 +4,6 @@ digraph CFG {
   1 [label="Block 1\lret _t0\l"];
   2 [label="Block 2\l_t3 <- 1\l_t4 <- _t0 - _t3\l_t5 <- fib(_t4)\l_t6 <- 2\l_t7 <- _t0 - _t6\l_t8 <- fib(_t7)\l_t9 <- _t5 + _t8\lret _t9\l"];
   -1 -> 2 [label=""];
-  0 -> 1 [label="True"];
   0 -> 2 [label="False"];
+  0 -> 1 [label="True"];
 }
diff --git a/fib_iterative.dot b/fib_iterative.dot
index 6b82fb7..596cc09 100644
--- a/fib_iterative.dot
+++ b/fib_iterative.dot
@@ -6,8 +6,8 @@ digraph CFG {
   3 [label="Block 3\l_t6 <- 1\li <- i + _t6\lujmp fib_iterative1\l"];
   4 [label="Block 4\lret b\l"];
   0 -> 1 [label=""];
-  1 -> 4 [label="False"];
   1 -> 2 [label="True"];
+  1 -> 4 [label="False"];
   2 -> 3 [label=""];
   3 -> 1 [label=""];
 }
diff --git a/fib_memoized.dot b/fib_memoized.dot
index 22c5fec..61be9a4 100644
--- a/fib_memoized.dot
+++ b/fib_memoized.dot
@@ -9,8 +9,8 @@ digraph CFG {
   6 [label="Block 6\lmemoization[_t0] <- res\lis_memoized[_t0] <- 1\lret res\l"];
   7 [label="Block 7\l_t5 <- 1\l_t6 <- _t0 == _t5\lcjmp _t6, fib_memoized4\lujmp fib_memoized5\l"];
   -1 -> 3 [label=""];
-  0 -> 2 [label="False"];
   0 -> 1 [label="True"];
+  0 -> 2 [label="False"];
   2 -> 4 [label="True"];
   2 -> 7 [label="False"];
   4 -> 6 [label=""];
diff --git a/fib_recursive.dot b/fib_recursive.dot
index 2caa00c..9839b2c 100644
--- a/fib_recursive.dot
+++ b/fib_recursive.dot
@@ -6,8 +6,8 @@ digraph CFG {
   3 [label="Block 3\lexit(-1)\l"];
   4 [label="Block 4\l_t3 <- 1\l_t4 <- _t0 == _t3\lcjmp _t4, fib_recursive1\lujmp fib_recursive2\l"];
   -1 -> 3 [label=""];
-  0 -> 4 [label="False"];
   0 -> 1 [label="True"];
-  4 -> 1 [label="True"];
+  0 -> 4 [label="False"];
   4 -> 2 [label="False"];
+  4 -> 1 [label="True"];
 }
diff --git a/foo.dot b/foo.dot
index 581d1a6..e6993d7 100644
--- a/foo.dot
+++ b/foo.dot
@@ -1,11 +1,4 @@
 digraph CFG {
   node [shape=box fontname="monospace"];
-  0 [label="Block 0\l_t1 <- 7\l_t2 <- get_int(_t1)\l_t3 <- 2\l_t4 <- get_int(_t3)\l_t5 <- get_bool(1)\l_t6 <- 0\l_c1 <- _t6\l_t7 <- _t2 + _t4\li <- _t7\lujmp foo1\l"];
-  1 [label="Block 1\l_t8 <- 2\l_t9 <- _t8 * _t7\l_t10 <- i < _t9\lcjmp _t10, foo2\lujmp foo3\l"];
-  2 [label="Block 2\l_t11 <- i * i\l_t12 <- _t11 * _t7\l_t13 <- _c1 + _t12\l_c1 <- _t13\l_t14 <- 1\l_t15 <- i + _t14\li <- _t15\lujmp foo1\l"];
-  3 [label="Block 3\lret i\l"];
-  0 -> 1 [label=""];
-  1 -> 3 [label="False"];
-  1 -> 2 [label="True"];
-  2 -> 1 [label=""];
+  0 [label="Block 0\l_t0 <- String(0, String)\l_t1 <- 42\lprintf(_t0, _t1)\lret\l"];
 }
diff --git a/foo.png b/foo.png
index f3d67ae..dbb3177 100644
Binary files a/foo.png and b/foo.png differ
diff --git a/fourth.dot b/fourth.dot
index 39e1cbd..e730786 100644
--- a/fourth.dot
+++ b/fourth.dot
@@ -1,4 +1,4 @@
 digraph CFG {
   node [shape=box fontname="monospace"];
-  0 [label="Block 0\l_t0 <- String(4, String)\lprintf(_t0)\l_t1 <- 4\lret _t1\l"];
+  0 [label="Block 0\l_t0 <- String(0, String)\lprintf(_t0)\l_t1 <- 4\lret _t1\l"];
 }
diff --git a/fourth.png b/fourth.png
index 4c88cd0..ac0f511 100644
Binary files a/fourth.png and b/fourth.png differ
diff --git a/func1.dot b/func1.dot
index 355160a..ce15488 100644
--- a/func1.dot
+++ b/func1.dot
@@ -9,7 +9,7 @@ digraph CFG {
   6 [label="Block 6\l_t10 <- A[i]\l_t11 <- 0\l_t12 <- _t10 != _t11\lcjmp _t12, func19\lujmp func110\l"];
   7 [label="Block 7\l_t15 <- 1\li <- i + _t15\lujmp func15\l"];
   8 [label="Block 8\l_t16 <- 0\li <- _t16\lujmp func111\l"];
-  9 [label="Block 9\l_t13 <- String(0, String)\lprintf(_t13)\lujmp func110\l"];
+  9 [label="Block 9\l_t13 <- String(3, String)\lprintf(_t13)\lujmp func110\l"];
   10 [label="Block 10\l_t14 <- a + i\lA[i] <- _t14\lujmp func17\l"];
   11 [label="Block 11\l_t17 <- 10\l_t18 <- i < _t17\lcjmp _t18, func112\lujmp func114\l"];
   12 [label="Block 12\l_t19 <- A[i]\lsum <- sum + _t19\lujmp func113\l"];
@@ -21,16 +21,16 @@ digraph CFG {
   2 -> 3 [label=""];
   3 -> 1 [label=""];
   4 -> 5 [label=""];
-  5 -> 6 [label="True"];
   5 -> 8 [label="False"];
-  6 -> 10 [label="False"];
+  5 -> 6 [label="True"];
   6 -> 9 [label="True"];
+  6 -> 10 [label="False"];
   7 -> 5 [label=""];
   8 -> 11 [label=""];
   9 -> 10 [label=""];
   10 -> 7 [label=""];
-  11 -> 14 [label="False"];
   11 -> 12 [label="True"];
+  11 -> 14 [label="False"];
   12 -> 13 [label=""];
   13 -> 11 [label=""];
 }
diff --git a/func1.png b/func1.png
index 0a11bd6..b6fb051 100644
Binary files a/func1.png and b/func1.png differ
diff --git a/get_bool.dot b/get_bool.dot
index 7e173f0..59bf749 100644
--- a/get_bool.dot
+++ b/get_bool.dot
@@ -1,4 +1,11 @@
 digraph CFG {
   node [shape=box fontname="monospace"];
-  0 [label="Block 0\l_t0 <- arg0\lret _t0\l"];
+  0 [label="Block 0\l_t0 <- arg0\lcjmp _t0, get_bool1\lujmp get_bool2\l"];
+  1 [label="Block 1\l_t2 <- 1\lnum <- _t2\lujmp get_bool3\l"];
+  2 [label="Block 2\l_t3 <- 2\lnum <- _t3\lujmp get_bool3\l"];
+  3 [label="Block 3\l_t4 <- String(1, String)\lprintf(_t4, num)\lret _t0\l"];
+  0 -> 2 [label="False"];
+  0 -> 1 [label="True"];
+  1 -> 3 [label=""];
+  2 -> 3 [label=""];
 }
diff --git a/get_bool.png b/get_bool.png
index 7b25052..94107a1 100644
Binary files a/get_bool.png and b/get_bool.png differ
diff --git a/get_char.dot b/get_char.dot
index 586f689..0002d72 100644
--- a/get_char.dot
+++ b/get_char.dot
@@ -1,30 +1,30 @@
 digraph CFG {
   node [shape=box fontname="monospace"];
   0 [label="Block 0\l_t0 <- arg0\l_t1 <- 7\l_t2 <- _t0 % _t1\l_t3 <- 0\l_t4 <- _t2 == _t3\lcjmp _t4, get_char1\lujmp get_char2\l"];
-  1 [label="Block 1\l_t5 <- String(1, String)\lprintf(_t5, _t0)\l_t6 <- 7\l_t7 <- _t0 % _t6\l_t8 <- 65 + _t7\lret _t8\l"];
+  1 [label="Block 1\l_t5 <- String(0, String)\lprintf(_t5, _t0)\l_t6 <- 7\l_t7 <- _t0 % _t6\l_t8 <- 65 + _t7\lret _t8\l"];
   2 [label="Block 2\l_t9 <- 7\l_t10 <- _t0 % _t9\l_t11 <- 1\l_t12 <- _t10 == _t11\lcjmp _t12, get_char4\lujmp get_char5\l"];
   3 [label="Block 3\lexit(-1)\l"];
-  4 [label="Block 4\l_t13 <- String(2, String)\lprintf(_t13, _t0)\l_t14 <- 7\l_t15 <- _t0 % _t14\l_t16 <- 66 - _t15\lret _t16\l"];
+  4 [label="Block 4\l_t13 <- String(1, String)\lprintf(_t13, _t0)\l_t14 <- 7\l_t15 <- _t0 % _t14\l_t16 <- 66 - _t15\lret _t16\l"];
   5 [label="Block 5\l_t17 <- 7\l_t18 <- _t0 % _t17\l_t19 <- 2\l_t20 <- _t18 == _t19\lcjmp _t20, get_char7\lujmp get_char8\l"];
   6 [label="Block 6\lujmp get_char3\l"];
-  7 [label="Block 7\l_t21 <- String(3, String)\lprintf(_t21, _t0)\l_t22 <- 7\l_t23 <- _t0 % _t22\l_t24 <- 67 + _t23\lret _t24\l"];
+  7 [label="Block 7\l_t21 <- String(2, String)\lprintf(_t21, _t0)\l_t22 <- 7\l_t23 <- _t0 % _t22\l_t24 <- 67 + _t23\lret _t24\l"];
   8 [label="Block 8\l_t25 <- 7\l_t26 <- _t0 % _t25\l_t27 <- 3\l_t28 <- _t26 == _t27\lcjmp _t28, get_char10\lujmp get_char11\l"];
   9 [label="Block 9\lujmp get_char6\l"];
-  10 [label="Block 10\l_t29 <- String(4, String)\lprintf(_t29, _t0)\l_t30 <- 7\l_t31 <- _t0 % _t30\l_t32 <- 68 - _t31\lret _t32\l"];
+  10 [label="Block 10\l_t29 <- String(3, String)\lprintf(_t29, _t0)\l_t30 <- 7\l_t31 <- _t0 % _t30\l_t32 <- 68 - _t31\lret _t32\l"];
   11 [label="Block 11\l_t33 <- 7\l_t34 <- _t0 % _t33\l_t35 <- 4\l_t36 <- _t34 == _t35\lcjmp _t36, get_char13\lujmp get_char14\l"];
   12 [label="Block 12\lujmp get_char9\l"];
-  13 [label="Block 13\l_t37 <- String(5, String)\lprintf(_t37, _t0)\l_t38 <- 7\l_t39 <- _t0 % _t38\l_t40 <- 69 + _t39\lret _t40\l"];
+  13 [label="Block 13\l_t37 <- String(4, String)\lprintf(_t37, _t0)\l_t38 <- 7\l_t39 <- _t0 % _t38\l_t40 <- 69 + _t39\lret _t40\l"];
   14 [label="Block 14\l_t41 <- 7\l_t42 <- _t0 % _t41\l_t43 <- 5\l_t44 <- _t42 == _t43\lcjmp _t44, get_char16\lujmp get_char17\l"];
   15 [label="Block 15\lujmp get_char12\l"];
-  16 [label="Block 16\l_t45 <- String(6, String)\lprintf(_t45, _t0)\l_t46 <- 7\l_t47 <- _t0 % _t46\l_t48 <- 70 + _t47\lret _t48\l"];
-  17 [label="Block 17\l_t49 <- String(7, String)\lprintf(_t49, _t0)\l_t50 <- 7\l_t51 <- _t0 % _t50\l_t52 <- 71 - _t51\lret _t52\l"];
+  16 [label="Block 16\l_t45 <- String(5, String)\lprintf(_t45, _t0)\l_t46 <- 7\l_t47 <- _t0 % _t46\l_t48 <- 70 + _t47\lret _t48\l"];
+  17 [label="Block 17\l_t49 <- String(6, String)\lprintf(_t49, _t0)\l_t50 <- 7\l_t51 <- _t0 % _t50\l_t52 <- 71 - _t51\lret _t52\l"];
   18 [label="Block 18\lujmp get_char15\l"];
   -1 -> 3 [label=""];
-  -1 -> 12 [label=""];
-  -1 -> 18 [label=""];
-  -1 -> 15 [label=""];
   -1 -> 9 [label=""];
+  -1 -> 12 [label=""];
   -1 -> 6 [label=""];
+  -1 -> 15 [label=""];
+  -1 -> 18 [label=""];
   0 -> 1 [label="True"];
   0 -> 2 [label="False"];
   2 -> 5 [label="False"];
@@ -32,14 +32,14 @@ digraph CFG {
   5 -> 7 [label="True"];
   5 -> 8 [label="False"];
   6 -> 3 [label=""];
-  8 -> 10 [label="True"];
   8 -> 11 [label="False"];
+  8 -> 10 [label="True"];
   9 -> 6 [label=""];
-  11 -> 13 [label="True"];
   11 -> 14 [label="False"];
+  11 -> 13 [label="True"];
   12 -> 9 [label=""];
-  14 -> 16 [label="True"];
   14 -> 17 [label="False"];
+  14 -> 16 [label="True"];
   15 -> 12 [label=""];
   18 -> 15 [label=""];
 }
diff --git a/get_char.png b/get_char.png
index ff0876b..0e3b384 100644
Binary files a/get_char.png and b/get_char.png differ
diff --git a/get_int.dot b/get_int.dot
index 7e173f0..6880959 100644
--- a/get_int.dot
+++ b/get_int.dot
@@ -1,4 +1,4 @@
 digraph CFG {
   node [shape=box fontname="monospace"];
-  0 [label="Block 0\l_t0 <- arg0\lret _t0\l"];
+  0 [label="Block 0\l_t0 <- arg0\l_t1 <- String(0, String)\lprintf(_t1, _t0)\lret _t0\l"];
 }
diff --git a/get_int.png b/get_int.png
index 7b25052..46ebe0a 100644
Binary files a/get_int.png and b/get_int.png differ
diff --git a/get_long.dot b/get_long.dot
index 84cc71d..82f639f 100644
--- a/get_long.dot
+++ b/get_long.dot
@@ -1,4 +1,4 @@
 digraph CFG {
   node [shape=box fontname="monospace"];
-  0 [label="Block 0\l_t0 <- arg0\l_t1 <- String(1, String)\lprintf(_t1, _t0)\lret _t0\l"];
+  0 [label="Block 0\l_t0 <- arg0\l_t1 <- String(2, String)\lprintf(_t1, _t0)\lret _t0\l"];
 }
diff --git a/get_long.png b/get_long.png
index d03dbb5..f13cc1e 100644
Binary files a/get_long.png and b/get_long.png differ
diff --git a/gurp.dot b/gurp.dot
index 337dcd7..823af76 100644
--- a/gurp.dot
+++ b/gurp.dot
@@ -1,9 +1,9 @@
 digraph CFG {
   node [shape=box fontname="monospace"];
-  0 [label="Block 0\l_t0 <- arg0\l_t1 <- arg1\l_t2 <- arg2\l_t3 <- arg3\l_t4 <- arg4\l_t5 <- arg5\l_t6 <- arg6\l_t7 <- arg7\l_t8 <- String(17, String)\lprintf(_t8)\l_t9 <- _t0 * _t1\l_t10 <- _t9 - _t4\l_t11 <- _t10 + _t3\l_t12 <- String(18, String)\lprintf(_t12)\l_t13 <- 1\l_t14 <- _t11 - _t13\l_t15 <- String(19, String)\lprintf(_t15)\l_t16 <- 3\l_t17 <- _t16 + _t14\l_t18 <- String(20, String)\lprintf(_t18)\l_t19 <- _t14 + _t17\l_t20 <- 2\l_t21 <- _t19 - _t20\l_t22 <- String(21, String)\lprintf(_t22)\l_t23 <- _t21 - _t17\ll <- _t23\l_t24 <- String(22, String)\lprintf(_t24)\l_t25 <- _t14 + _t21\l_t26 <- String(23, String)\lprintf(_t26, _t21, _t25)\l_t27 <- _t21 - _t25\l_t28 <- _t17 - _t27\l_t29 <- _t14 + _t28\l_t30 <- String(24, String)\lprintf(_t30)\l_t31 <- 50\l_t32 <- _t0 > _t31\lcjmp _t32, gurp1\lujmp gurp2\l"];
+  0 [label="Block 0\l_t0 <- arg0\l_t1 <- arg1\l_t2 <- arg2\l_t3 <- arg3\l_t4 <- arg4\l_t5 <- arg5\l_t6 <- arg6\l_t7 <- arg7\l_t8 <- String(0, String)\lprintf(_t8)\l_t9 <- _t0 * _t1\l_t10 <- _t9 - _t4\l_t11 <- _t10 + _t3\l_t12 <- String(1, String)\lprintf(_t12)\l_t13 <- 1\l_t14 <- _t11 - _t13\l_t15 <- String(2, String)\lprintf(_t15)\l_t16 <- 3\l_t17 <- _t16 + _t14\l_t18 <- String(3, String)\lprintf(_t18)\l_t19 <- _t14 + _t17\l_t20 <- 2\l_t21 <- _t19 - _t20\l_t22 <- String(4, String)\lprintf(_t22)\l_t23 <- _t21 - _t17\ll <- _t23\l_t24 <- String(5, String)\lprintf(_t24)\l_t25 <- _t14 + _t21\l_t26 <- String(6, String)\lprintf(_t26, _t21, _t25)\l_t27 <- _t21 - _t25\l_t28 <- _t17 - _t27\l_t29 <- _t14 + _t28\l_t30 <- String(7, String)\lprintf(_t30)\l_t31 <- 50\l_t32 <- _t0 > _t31\lcjmp _t32, gurp1\lujmp gurp2\l"];
   1 [label="Block 1\l_t33 <- _t0 + _t1\l_t34 <- _t33 + _t23\lo <- _t34\l_t35 <- _t23 - _t25\lj <- _t35\lcjmp _t2, gurp4\lujmp gurp5\l"];
   2 [label="Block 2\l_t48 <- _t1 - _t0\lo <- _t48\l_t49 <- _t23 + _t25\lj <- _t49\lcjmp _t6, gurp10\lujmp gurp11\l"];
-  3 [label="Block 3\l_t64 <- String(25, String)\lprintf(_t64)\l_t65 <- _t4 + l\l_t66 <- _t65 - _t14\l_t67 <- _t66 + j\l_t68 <- _t67 + o\l_t69 <- _t66 - _t68\l_t70 <- _t69 + _t25\l_t71 <- _t70 - j\l_t72 <- l + _t68\l_t73 <- q + _t5\l_t74 <- _t66 - _t73\l_t75 <- _t72 + _t74\lcjmp _t6, gurp13\lujmp gurp14\l"];
+  3 [label="Block 3\l_t64 <- String(8, String)\lprintf(_t64)\l_t65 <- _t4 + l\l_t66 <- _t65 - _t14\l_t67 <- _t66 + j\l_t68 <- _t67 + o\l_t69 <- _t66 - _t68\l_t70 <- _t69 + _t25\l_t71 <- _t70 - j\l_t72 <- l + _t68\l_t73 <- q + _t5\l_t74 <- _t66 - _t73\l_t75 <- _t72 + _t74\lcjmp _t6, gurp13\lujmp gurp14\l"];
   4 [label="Block 4\l_t36 <- _t4 + _t5\l_t37 <- _t7 + _t29\l_t38 <- _t36 - _t37\ll <- _t38\l_t39 <- 1\l_t40 <- _t5 - _t39\l_t5 <- _t40\l_t41 <- 3\l_t42 <- _t34 - _t41\lq <- _t42\lujmp gurp6\l"];
   5 [label="Block 5\l_t43 <- 3\l_t44 <- _t34 + _t43\lq <- _t44\lujmp gurp6\l"];
   6 [label="Block 6\lcjmp _t2, gurp7\lujmp gurp9\l"];
@@ -15,28 +15,28 @@ digraph CFG {
   12 [label="Block 12\lujmp gurp3\l"];
   13 [label="Block 13\l_t76 <- 3\lw <- _t76\lujmp gurp15\l"];
   14 [label="Block 14\l_t77 <- -1\lw <- _t77\lujmp gurp15\l"];
-  15 [label="Block 15\l_t78 <- String(26, String)\lprintf(_t78)\l_t79 <- 99\l_t80 <- _t79 + _t29\l_t81 <- _t80 - _t4\l_t82 <- _t81 + _t66\l_t83 <- _t4 - s\l_t84 <- _t83 + _t29\l_t85 <- _t84 + o\l_t86 <- _t85 + _t25\l_t87 <- _t86 - _t4\l_t88 <- _t87 + _t1\l_t89 <- _t88 - _t1\l_t90 <- _t89 + _t4\l_t91 <- _t68 + _t66\l_t92 <- _t91 - _t29\l_t93 <- _t92 - _t66\l_t94 <- _t7 + _t5\l_t95 <- _t94 + _t3\l_t96 <- _t95 - _t4\l_t97 <- -1\l_t98 <- w == _t97\lcjmp _t98, gurp16\lujmp gurp17\l"];
+  15 [label="Block 15\l_t78 <- String(9, String)\lprintf(_t78)\l_t79 <- 99\l_t80 <- _t79 + _t29\l_t81 <- _t80 - _t4\l_t82 <- _t81 + _t66\l_t83 <- _t4 - s\l_t84 <- _t83 + _t29\l_t85 <- _t84 + o\l_t86 <- _t85 + _t25\l_t87 <- _t86 - _t4\l_t88 <- _t87 + _t1\l_t89 <- _t88 - _t1\l_t90 <- _t89 + _t4\l_t91 <- _t68 + _t66\l_t92 <- _t91 - _t29\l_t93 <- _t92 - _t66\l_t94 <- _t7 + _t5\l_t95 <- _t94 + _t3\l_t96 <- _t95 - _t4\l_t97 <- -1\l_t98 <- w == _t97\lcjmp _t98, gurp16\lujmp gurp17\l"];
   16 [label="Block 16\l_t99 <- _t93 + _t96\la3 <- _t99\lujmp gurp18\l"];
   17 [label="Block 17\l_t100 <- _t96 - _t93\la3 <- _t100\lujmp gurp18\l"];
-  18 [label="Block 18\l_t101 <- _t82 - _t75\l_t102 <- _t101 + w\l_t103 <- 42\l_t104 <- _t103 - _t102\l_t105 <- 0\l_t106 <- String(27, String)\lprintf(_t106)\l_t107 <- String(28, String)\l_t108 <- _t0 - _t14\l_t109 <- _t108 + _t3\l_t110 <- _t109 + j\l_t111 <- _t110 - _t4\l_t112 <- _t111 - _t5\l_t113 <- _t112 + l\l_t114 <- _t113 + _t25\lprintf(_t107, _t114)\l_t115 <- String(29, String)\lprintf(_t115)\l_t116 <- String(30, String)\l_t117 <- _t1 - _t21\l_t118 <- _t117 - _t7\l_t119 <- _t118 + _t29\l_t120 <- _t119 + o\l_t121 <- _t120 + q\l_t122 <- _t121 - _t71\l_t123 <- _t122 - _t93\lprintf(_t116, _t123)\l_t124 <- String(31, String)\lprintf(_t124)\l_t125 <- String(32, String)\l_t126 <- _t96 + s\l_t127 <- _t126 - _t66\l_t128 <- _t127 + _t102\l_t129 <- _t128 - _t90\l_t130 <- _t129 + _t105\l_t131 <- _t130 + w\l_t132 <- _t131 - a3\lprintf(_t125, _t132)\l_t133 <- String(33, String)\lprintf(_t133)\lcjmp _t2, gurp19\lujmp gurp20\l"];
-  19 [label="Block 19\l_t134 <- String(34, String)\l_t135 <- _t104 + _t82\l_t136 <- _t135 - _t75\l_t137 <- _t136 + _t68\lprintf(_t134, _t137)\lujmp gurp21\l"];
+  18 [label="Block 18\l_t101 <- _t82 - _t75\l_t102 <- _t101 + w\l_t103 <- 42\l_t104 <- _t103 - _t102\l_t105 <- 0\l_t106 <- String(10, String)\lprintf(_t106)\l_t107 <- String(11, String)\l_t108 <- _t0 - _t14\l_t109 <- _t108 + _t3\l_t110 <- _t109 + j\l_t111 <- _t110 - _t4\l_t112 <- _t111 - _t5\l_t113 <- _t112 + l\l_t114 <- _t113 + _t25\lprintf(_t107, _t114)\l_t115 <- String(12, String)\lprintf(_t115)\l_t116 <- String(13, String)\l_t117 <- _t1 - _t21\l_t118 <- _t117 - _t7\l_t119 <- _t118 + _t29\l_t120 <- _t119 + o\l_t121 <- _t120 + q\l_t122 <- _t121 - _t71\l_t123 <- _t122 - _t93\lprintf(_t116, _t123)\l_t124 <- String(14, String)\lprintf(_t124)\l_t125 <- String(15, String)\l_t126 <- _t96 + s\l_t127 <- _t126 - _t66\l_t128 <- _t127 + _t102\l_t129 <- _t128 - _t90\l_t130 <- _t129 + _t105\l_t131 <- _t130 + w\l_t132 <- _t131 - a3\lprintf(_t125, _t132)\l_t133 <- String(16, String)\lprintf(_t133)\lcjmp _t2, gurp19\lujmp gurp20\l"];
+  19 [label="Block 19\l_t134 <- String(17, String)\l_t135 <- _t104 + _t82\l_t136 <- _t135 - _t75\l_t137 <- _t136 + _t68\lprintf(_t134, _t137)\lujmp gurp21\l"];
   20 [label="Block 20\lcjmp _t6, gurp22\lujmp gurp23\l"];
-  21 [label="Block 21\l_t146 <- String(37, String)\lprintf(_t146)\l_t147 <- _t0 - _t1\l_t148 <- _t147 + _t3\l_t149 <- _t148 + _t4\l_t150 <- _t149 + _t5\l_t151 <- _t150 + _t7\l_t152 <- _t151 + _t14\l_t153 <- _t152 + j\l_t154 <- _t153 + _t21\l_t155 <- _t154 + l\l_t156 <- _t155 + _t25\l_t157 <- _t156 - _t29\l_t158 <- _t157 + o\l_t159 <- _t158 - _t104\l_t160 <- _t159 + q\l_t161 <- _t160 - _t71\l_t162 <- _t161 + s\l_t163 <- _t162 - _t66\l_t164 <- _t163 + _t82\l_t165 <- _t164 - _t75\l_t166 <- _t165 + w\l_t167 <- _t166 - _t105\l_t168 <- _t167 + _t90\l_t169 <- _t168 - _t102\l_t170 <- _t169 + _t68\l_t171 <- _t170 + _t93\l_t172 <- _t171 + a3\l_t173 <- _t172 - _t96\lret _t173\l"];
-  22 [label="Block 22\l_t138 <- String(35, String)\l_t140 <- _t101 + _t68\l_t141 <- _t140 - _t104\lprintf(_t138, _t141)\lujmp gurp24\l"];
-  23 [label="Block 23\l_t142 <- String(36, String)\l_t143 <- _t68 + _t75\l_t144 <- _t143 + _t104\l_t145 <- _t144 - _t82\lprintf(_t142, _t145)\lujmp gurp24\l"];
+  21 [label="Block 21\l_t146 <- String(20, String)\lprintf(_t146)\l_t147 <- _t0 - _t1\l_t148 <- _t147 + _t3\l_t149 <- _t148 + _t4\l_t150 <- _t149 + _t5\l_t151 <- _t150 + _t7\l_t152 <- _t151 + _t14\l_t153 <- _t152 + j\l_t154 <- _t153 + _t21\l_t155 <- _t154 + l\l_t156 <- _t155 + _t25\l_t157 <- _t156 - _t29\l_t158 <- _t157 + o\l_t159 <- _t158 - _t104\l_t160 <- _t159 + q\l_t161 <- _t160 - _t71\l_t162 <- _t161 + s\l_t163 <- _t162 - _t66\l_t164 <- _t163 + _t82\l_t165 <- _t164 - _t75\l_t166 <- _t165 + w\l_t167 <- _t166 - _t105\l_t168 <- _t167 + _t90\l_t169 <- _t168 - _t102\l_t170 <- _t169 + _t68\l_t171 <- _t170 + _t93\l_t172 <- _t171 + a3\l_t173 <- _t172 - _t96\lret _t173\l"];
+  22 [label="Block 22\l_t138 <- String(18, String)\l_t140 <- _t101 + _t68\l_t141 <- _t140 - _t104\lprintf(_t138, _t141)\lujmp gurp24\l"];
+  23 [label="Block 23\l_t142 <- String(19, String)\l_t143 <- _t68 + _t75\l_t144 <- _t143 + _t104\l_t145 <- _t144 - _t82\lprintf(_t142, _t145)\lujmp gurp24\l"];
   24 [label="Block 24\lujmp gurp21\l"];
-  0 -> 1 [label="True"];
   0 -> 2 [label="False"];
-  1 -> 5 [label="False"];
+  0 -> 1 [label="True"];
   1 -> 4 [label="True"];
+  1 -> 5 [label="False"];
   2 -> 10 [label="True"];
   2 -> 11 [label="False"];
-  3 -> 14 [label="False"];
   3 -> 13 [label="True"];
+  3 -> 14 [label="False"];
   4 -> 6 [label=""];
   5 -> 6 [label=""];
-  6 -> 7 [label="True"];
   6 -> 9 [label="False"];
+  6 -> 7 [label="True"];
   7 -> 8 [label=""];
   8 -> 3 [label=""];
   9 -> 7 [label="True"];
@@ -46,15 +46,15 @@ digraph CFG {
   12 -> 3 [label=""];
   13 -> 15 [label=""];
   14 -> 15 [label=""];
-  15 -> 17 [label="False"];
   15 -> 16 [label="True"];
+  15 -> 17 [label="False"];
   16 -> 18 [label=""];
   17 -> 18 [label=""];
-  18 -> 19 [label="True"];
   18 -> 20 [label="False"];
+  18 -> 19 [label="True"];
   19 -> 21 [label=""];
-  20 -> 23 [label="False"];
   20 -> 22 [label="True"];
+  20 -> 23 [label="False"];
   22 -> 24 [label=""];
   23 -> 24 [label=""];
   24 -> 21 [label=""];
diff --git a/gurp.png b/gurp.png
index 247405f..e2f2d18 100644
Binary files a/gurp.png and b/gurp.png differ
diff --git a/insertionSort.dot b/insertionSort.dot
index 7706806..1decd62 100644
--- a/insertionSort.dot
+++ b/insertionSort.dot
@@ -1,9 +1,9 @@
 digraph CFG {
   node [shape=box fontname="monospace"];
-  0 [label="Block 0\l_t0 <- 1\l_i1 <- _t0\lujmp insertionSort1\l"];
-  1 [label="Block 1\l_t1 <- _i1 < n\lcjmp _t1, insertionSort2\lujmp insertionSort4\l"];
-  2 [label="Block 2\l_t2 <- a[_i1]\lkey <- _t2\l_t3 <- 1\l_t4 <- _i1 - _t3\lj <- _t4\lujmp insertionSort5\l"];
-  3 [label="Block 3\l_t15 <- 1\l_i1 <- _i1 + _t15\lujmp insertionSort1\l"];
+  0 [label="Block 0\l_t0 <- 1\li <- _t0\lujmp insertionSort1\l"];
+  1 [label="Block 1\l_t1 <- i < n\lcjmp _t1, insertionSort2\lujmp insertionSort4\l"];
+  2 [label="Block 2\l_t2 <- a[i]\lkey <- _t2\l_t3 <- 1\l_t4 <- i - _t3\lj <- _t4\lujmp insertionSort5\l"];
+  3 [label="Block 3\l_t15 <- 1\li <- i + _t15\lujmp insertionSort1\l"];
   4 [label="Block 4\l\l"];
   5 [label="Block 5\l_t5 <- 0\l_t6 <- j >= _t5\lcjmp _t6, insertionSort8\lujmp insertionSort7\l"];
   6 [label="Block 6\l_t9 <- a[j]\l_t10 <- 1\l_t11 <- j + _t10\la[_t11] <- _t9\l_t12 <- 1\lj <- j - _t12\lujmp insertionSort5\l"];
@@ -14,8 +14,8 @@ digraph CFG {
   1 -> 2 [label="True"];
   2 -> 5 [label=""];
   3 -> 1 [label=""];
-  5 -> 8 [label="True"];
   5 -> 7 [label="False"];
+  5 -> 8 [label="True"];
   6 -> 5 [label=""];
   7 -> 3 [label=""];
   8 -> 7 [label="False"];
diff --git a/insertionSort.png b/insertionSort.png
index d0a172f..1cc6d42 100644
Binary files a/insertionSort.png and b/insertionSort.png differ
diff --git a/loops.dot b/loops.dot
index 22bc433..b350beb 100644
--- a/loops.dot
+++ b/loops.dot
@@ -13,8 +13,8 @@ digraph CFG {
   1 -> 4 [label="False"];
   2 -> 5 [label=""];
   3 -> 1 [label=""];
-  5 -> 6 [label="True"];
   5 -> 7 [label="False"];
+  5 -> 6 [label="True"];
   6 -> 5 [label=""];
   7 -> 3 [label=""];
 }
diff --git a/m2.dot b/m2.dot
index 1a08083..3312dde 100644
--- a/m2.dot
+++ b/m2.dot
@@ -1,7 +1,7 @@
 digraph CFG {
   node [shape=box fontname="monospace"];
   0 [label="Block 0\l_t0 <- arg0\l_t1 <- arg1\l_t2 <- arg2\l_t3 <- arg3\l_t4 <- arg4\l_t5 <- arg5\l_t6 <- arg6\l_t7 <- arg7\l_t8 <- arg8\l_t13 <- arg13\l_t14 <- 1\l_t15 <- _t0 != _t14\lcjmp _t15, m21\lujmp m211\l"];
-  1 [label="Block 1\l_t34 <- String(4, String)\lprintf(_t34)\lujmp m22\l"];
+  1 [label="Block 1\l_t34 <- String(0, String)\lprintf(_t34)\lujmp m22\l"];
   2 [label="Block 2\l_t35 <- 1\lret _t35\l"];
   3 [label="Block 3\l_t32 <- 10\l_t33 <- _t13 != _t32\lcjmp _t33, m21\lujmp m22\l"];
   4 [label="Block 4\l_t30 <- 9\l_t31 <- _t8 != _t30\lcjmp _t31, m21\lujmp m23\l"];
@@ -15,22 +15,22 @@ digraph CFG {
   0 -> 1 [label="True"];
   0 -> 11 [label="False"];
   1 -> 2 [label=""];
-  3 -> 2 [label="False"];
   3 -> 1 [label="True"];
+  3 -> 2 [label="False"];
   4 -> 1 [label="True"];
   4 -> 3 [label="False"];
-  5 -> 1 [label="True"];
   5 -> 4 [label="False"];
+  5 -> 1 [label="True"];
   6 -> 1 [label="True"];
   6 -> 5 [label="False"];
-  7 -> 6 [label="False"];
   7 -> 1 [label="True"];
+  7 -> 6 [label="False"];
   8 -> 7 [label="False"];
   8 -> 1 [label="True"];
-  9 -> 1 [label="True"];
   9 -> 8 [label="False"];
-  10 -> 1 [label="True"];
+  9 -> 1 [label="True"];
   10 -> 9 [label="False"];
+  10 -> 1 [label="True"];
   11 -> 1 [label="True"];
   11 -> 10 [label="False"];
 }
diff --git a/m2.png b/m2.png
index e38fc16..0113f0e 100644
Binary files a/m2.png and b/m2.png differ
diff --git a/main.dot b/main.dot
index 4fa4496..3e2e123 100644
--- a/main.dot
+++ b/main.dot
@@ -1,4 +1,13 @@
 digraph CFG {
   node [shape=box fontname="monospace"];
-  0 [label="Block 0\l_t0 <- foo()\l_t1 <- String(0, String)\lprintf(_t1, _t0)\l_t2 <- 0\lret _t2\l"];
+  0 [label="Block 0\l_t1 <- 0\lc <- _t1\l_t2 <- 0\li <- _t2\lujmp main1\l"];
+  1 [label="Block 1\l_t3 <- 10\l_t4 <- i < _t3\lcjmp _t4, main2\lujmp main4\l"];
+  2 [label="Block 2\l_t5 <- c + i\lc <- _t5\lujmp main3\l"];
+  3 [label="Block 3\l_t6 <- 1\li <- i + _t6\lujmp main1\l"];
+  4 [label="Block 4\l_t7 <- 0\lret _t7\l"];
+  0 -> 1 [label=""];
+  1 -> 2 [label="True"];
+  1 -> 4 [label="False"];
+  2 -> 3 [label=""];
+  3 -> 1 [label=""];
 }
diff --git a/main.png b/main.png
index 7582c7e..9b855fe 100644
Binary files a/main.png and b/main.png differ
diff --git a/main_x.dot b/main_x.dot
index 04257c7..ea4079a 100644
--- a/main_x.dot
+++ b/main_x.dot
@@ -7,10 +7,10 @@ digraph CFG {
   4 [label="Block 4\l_t8 <- String(2, String)\l_t9 <- String(3, String)\l_t10 <- strcmp(_t8, _t9)\l_t11 <- String(4, String)\l_t12 <- printf(_t11)\l_t13 <- _t10 * _t12\lexit(_t13)\lmain_x()\lujmp main_x5\l"];
   5 [label="Block 5\l\l"];
   -1 -> 3 [label=""];
-  0 -> 2 [label="False"];
   0 -> 1 [label="True"];
+  0 -> 2 [label="False"];
   1 -> 3 [label=""];
-  3 -> 5 [label="False"];
   3 -> 4 [label="True"];
+  3 -> 5 [label="False"];
   4 -> 5 [label=""];
 }
diff --git a/not2.dot b/not2.dot
index 4861324..c5298fc 100644
--- a/not2.dot
+++ b/not2.dot
@@ -4,8 +4,8 @@ digraph CFG {
   1 [label="Block 1\lb <- 0\lujmp not23\l"];
   2 [label="Block 2\lb <- 1\lujmp not23\l"];
   3 [label="Block 3\lret b\l"];
-  0 -> 1 [label="True"];
   0 -> 2 [label="False"];
+  0 -> 1 [label="True"];
   1 -> 3 [label=""];
   2 -> 3 [label=""];
 }
diff --git a/opt.html b/opt.html
index e274c57..7b74d86 100644
--- a/opt.html
+++ b/opt.html
@@ -1,6 +1,3 @@
 <!DOCTYPE html><html><head><meta charset="UTF-8"><title>CFG Viewer</title></head><body><h1>CFG Visualizations</h1>
 <h2>main</h2><img src="main.png" style="border:1px solid #ccc; max-width:100%;"><br><br>
-<h2>get_int</h2><img src="get_int.png" style="border:1px solid #ccc; max-width:100%;"><br><br>
-<h2>get_bool</h2><img src="get_bool.png" style="border:1px solid #ccc; max-width:100%;"><br><br>
-<h2>foo</h2><img src="foo.png" style="border:1px solid #ccc; max-width:100%;"><br><br>
 </body></html>
diff --git a/out.o b/out.o
index 468094e..545d794 100755
Binary files a/out.o and b/out.o differ
diff --git a/out.s b/out.s
index 872385c..10c04cc 100644
--- a/out.s
+++ b/out.s
@@ -1,260 +1,22 @@
-<<<<<<< HEAD
-Replacing j with _t1
-Replacing j with _t1
-Replacing sum with _t0
-Replacing j with _t1
-Replacing sum with _t0
-Replacing j with _t1
-Replacing sum with _t0
-Replacing j with _t11
-Replacing i with _t10
-Replacing j with _t11
-Replacing i with _t10
-Replacing i with _t10
 
 str0:
-    .string "j is %d\n"
-str1:
-    .string "sum is %d\n"
-str2:
-    .string "ERROR: for loop is bad (1)\n"
-str3:
-    .string "hi\n"
-str4:
-    .string "ERROR: for loop is bad (2)\n"
-str5:
     .string "%d\n"
-str6:
-    .string "ERROR: true branch is bad (2)\n"
-str7:
-    .string "ERROR: else branch is bad (2)\n"
-str8:
-    .string "control flow OK if no previous output\n"
-.globl main
-main:
-    pushq %rbp
-    movq %rsp, %rbp
-    subq $176, %rsp
-main0:
-    movl $0, -16(%rbp)
-    movl -16(%rbp), %eax
-    movl %eax, -8(%rbp)
-    movl $0, -20(%rbp)
-    movl -20(%rbp), %eax
-    movl %eax, -12(%rbp)
-    jmp main1
-main1:
-    movl $10, -24(%rbp)
-    movl -20(%rbp), %eax
-    cmpl -24(%rbp), %eax
-    setl %al
-    movzbq %al, %rax
-    movl %eax, -28(%rbp)
-    movl -28(%rbp), %eax
-    cmpl $0, %eax
-    jne main2
-
-    jmp main4
-main2:
-    leaq str0(%rip), %rax
-    movq %rax, -36(%rbp)
-    movq -36(%rbp), %rdi
-    movl -20(%rbp), %esi
-    movq $0, %rax
-    call printf
-    movl -16(%rbp), %eax
-    addl -20(%rbp), %eax
-    movl %eax, -8(%rbp)
-    leaq str1(%rip), %rax
-    movq %rax, -44(%rbp)
-    movq -44(%rbp), %rdi
-    movl -16(%rbp), %esi
-    movq $0, %rax
-    call printf
-    jmp main3
-main3:
-    movl $1, -48(%rbp)
-    movl -20(%rbp), %eax
-    addl -48(%rbp), %eax
-    movl %eax, -12(%rbp)
-    jmp main1
-main4:
-    movl $45, -52(%rbp)
-    movl -16(%rbp), %eax
-    cmpl -52(%rbp), %eax
-    setne %al
-    movzbq %al, %rax
-    movl %eax, -56(%rbp)
-    movl -56(%rbp), %eax
-    cmpl $0, %eax
-    jne main5
-
-    jmp main6
-main5:
-    leaq str2(%rip), %rax
-    movq %rax, -64(%rbp)
-    movq -64(%rbp), %rdi
-    movq $0, %rax
-    call printf
-    jmp main6
-main6:
-    movl $11, -68(%rbp)
-    movl -68(%rbp), %eax
-    movl %eax, -4(%rbp)
-    movl $10, -72(%rbp)
-    movl -72(%rbp), %eax
-    movl %eax, -12(%rbp)
-    jmp main7
-main7:
-    movl $0, -76(%rbp)
-    movl -72(%rbp), %eax
-    cmpl -76(%rbp), %eax
-    setl %al
-    movzbq %al, %rax
-    movl %eax, -80(%rbp)
-    movl -80(%rbp), %eax
-    cmpl $0, %eax
-    jne main8
-
-    jmp main10
-main8:
-    leaq str3(%rip), %rax
-    movq %rax, -88(%rbp)
-    movq -88(%rbp), %rdi
-    movq $0, %rax
-    call printf
-    movl $1, -92(%rbp)
-    movl -68(%rbp), %eax
-    addl -92(%rbp), %eax
-    movl %eax, -4(%rbp)
-    jmp main9
-main9:
-    movl $1, -96(%rbp)
-    movl -72(%rbp), %eax
-    addl -96(%rbp), %eax
-    movl %eax, -12(%rbp)
-    jmp main7
-main10:
-    movl $11, -100(%rbp)
-    movl -68(%rbp), %eax
-    cmpl -100(%rbp), %eax
-    setne %al
-    movzbq %al, %rax
-    movl %eax, -104(%rbp)
-    movl -104(%rbp), %eax
-    cmpl $0, %eax
-    jne main11
-
-    jmp main12
-main11:
-    leaq str4(%rip), %rax
-    movq %rax, -112(%rbp)
-    movq -112(%rbp), %rdi
-    movq $0, %rax
-    call printf
-    leaq str5(%rip), %rax
-    movq %rax, -120(%rbp)
-    movq -120(%rbp), %rdi
-    movl -68(%rbp), %esi
-    movq $0, %rax
-    call printf
-    jmp main12
-main12:
-    movl $1, -124(%rbp)
-    movl $2, -128(%rbp)
-    movl -124(%rbp), %eax
-    cmpl -128(%rbp), %eax
-    setg %al
-    movzbq %al, %rax
-    movl %eax, -132(%rbp)
-    movl -132(%rbp), %eax
-    cmpl $0, %eax
-    jne main13
-
-    jmp main14
-main13:
-    leaq str6(%rip), %rax
-    movq %rax, -140(%rbp)
-    movq -140(%rbp), %rdi
-    movq $0, %rax
-    call printf
-    jmp main15
-main14:
-    jmp main15
-main15:
-    movl $1, -144(%rbp)
-    movl $2, -148(%rbp)
-    movl -144(%rbp), %eax
-    cmpl -148(%rbp), %eax
-    setl %al
-    movzbq %al, %rax
-    movl %eax, -152(%rbp)
-    movl -152(%rbp), %eax
-    cmpl $0, %eax
-    jne main16
-
-    jmp main17
-main16:
-    jmp main18
-main17:
-    leaq str7(%rip), %rax
-    movq %rax, -160(%rbp)
-    movq -160(%rbp), %rdi
-    movq $0, %rax
-    call printf
-    jmp main18
-main18:
-    leaq str8(%rip), %rax
-    movq %rax, -168(%rbp)
-    movq -168(%rbp), %rdi
-    movq $0, %rax
-    call printf
-    movl $0, -172(%rbp)
-    movl -172(%rbp), %eax
-    movq %rbp, %rsp
-    popq %rbp
-    ret
-
-||||||| 3014bf8
-=======
-
-
-str0:
-    .string "%ld\n"
 str1:
-    .string "%d\n"
+    .string "%ld\n"
 get_int:
     pushq %rbp
     movq %rsp, %rbp
     subq $16, %rsp
 get_int0:
-    movq %rdi, %rax
-    movq %rax, -8(%rbp)
-    leaq str1(%rip), %rax
-    movq %rax, -16(%rbp)
-    movq -16(%rbp), %rdi
-    movq -8(%rbp), %rsi
-    movq $0, %rax
-    call printf
-    movq -8(%rbp), %rax
-    movq %rbp, %rsp
-    popq %rbp
-    ret
-
-get_long:
-    pushq %rbp
-    movq %rsp, %rbp
-    subq $16, %rsp
-get_long0:
-    movq %rdi, %rax
-    movq %rax, -8(%rbp)
+    movl %edi, %eax
+    movl %eax, -4(%rbp)
     leaq str0(%rip), %rax
-    movq %rax, -16(%rbp)
-    movq -16(%rbp), %rdi
-    movq -8(%rbp), %rsi
+    movq %rax, -12(%rbp)
+    movq -12(%rbp), %rdi
+    movl -4(%rbp), %esi
     movq $0, %rax
     call printf
-    movq -8(%rbp), %rax
+    movl -4(%rbp), %eax
     movq %rbp, %rsp
     popq %rbp
     ret
@@ -263,403 +25,328 @@ get_long0:
 main:
     pushq %rbp
     movq %rsp, %rbp
-    subq $688, %rsp
+    subq $528, %rsp
 main0:
-    movq $10, %rbx
-    movq $0, %rax
-    shlq $32, %rax
-    orq %rax, %rbx
-    movq %rbx, -32(%rbp)
-    movq -32(%rbp), %rax
-    movq %rax, -8(%rbp)
+    movl $10, -20(%rbp)
+    movl -20(%rbp), %eax
+    movl %eax, -4(%rbp)
     movq $4284967296, %rbx
     movq $4294967295, %rax
     shlq $32, %rax
     orq %rax, %rbx
-    movq %rbx, -40(%rbp)
-    movq -40(%rbp), %rax
-    movq %rax, -16(%rbp)
-    movq $4294967292, %rbx
-    movq $4294967295, %rax
-    shlq $32, %rax
-    orq %rax, %rbx
-    movq %rbx, -48(%rbp)
-    movq -48(%rbp), %rax
-    movq %rax, -24(%rbp)
+    movq %rbx, -28(%rbp)
+    movq -28(%rbp), %rax
+    movq %rax, -12(%rbp)
+    movl $-4, -32(%rbp)
+    movl -32(%rbp), %eax
+    movl %eax, -16(%rbp)
     jmp main1
 main1:
-    movq $6, %rbx
-    movq $0, %rax
-    shlq $32, %rax
-    orq %rax, %rbx
-    movq %rbx, -56(%rbp)
-    movq -24(%rbp), %rax
-    cmpq -56(%rbp), %rax
+    movl $6, -36(%rbp)
+    movl -16(%rbp), %eax
+    cmpl -36(%rbp), %eax
     setl %al
     movzbq %al, %rax
-    movq %rax, -64(%rbp)
-    movq -64(%rbp), %rax
-    cmpq $0, %rax
+    movl %eax, -40(%rbp)
+    movl -40(%rbp), %eax
+    cmpl $0, %eax
     jne main2
 
     jmp main4
 main2:
-    movq -24(%rbp), %rdi
+    movl -16(%rbp), %edi
     movq $0, %rax
     call get_int
-    movq %rax, -72(%rbp)
-    movq -8(%rbp), %rax
-    addq -72(%rbp), %rax
-    movq %rax, -8(%rbp)
-    movq -24(%rbp), %rdi
+    movl %eax, -44(%rbp)
+    movl -4(%rbp), %eax
+    addl -44(%rbp), %eax
+    movl %eax, -4(%rbp)
+    movl -16(%rbp), %edi
     movq $0, %rax
     call get_int
-    movq %rax, -80(%rbp)
-    movq $3, %rbx
-    movq $0, %rax
-    shlq $32, %rax
-    orq %rax, %rbx
-    movq %rbx, -88(%rbp)
-    movq -8(%rbp), %rax
-    subq -88(%rbp), %rax
-    movq %rax, -96(%rbp)
-    movq -96(%rbp), %rdi
+    movl %eax, -48(%rbp)
+    movl $3, -52(%rbp)
+    movl -4(%rbp), %eax
+    subl -52(%rbp), %eax
+    movl %eax, -56(%rbp)
+    movl -56(%rbp), %edi
     movq $0, %rax
     call get_int
-    movq %rax, -104(%rbp)
-    movq -80(%rbp), %rax
-    subq -104(%rbp), %rax
-    movq %rax, -112(%rbp)
-    movq -8(%rbp), %rax
-    subq -112(%rbp), %rax
-    movq %rax, -8(%rbp)
-    movq -24(%rbp), %rdi
+    movl %eax, -60(%rbp)
+    movl -48(%rbp), %eax
+    subl -60(%rbp), %eax
+    movl %eax, -64(%rbp)
+    movl -4(%rbp), %eax
+    subl -64(%rbp), %eax
+    movl %eax, -4(%rbp)
+    movl -16(%rbp), %edi
     movq $0, %rax
     call get_int
-    movq %rax, -120(%rbp)
-    movq $5, %rbx
-    movq $0, %rax
-    shlq $32, %rax
-    orq %rax, %rbx
-    movq %rbx, -128(%rbp)
-    movq -8(%rbp), %rax
-    imul -128(%rbp), %rax
-    movq %rax, -136(%rbp)
-    movq -136(%rbp), %rdi
+    movl %eax, -68(%rbp)
+    movl $5, -72(%rbp)
+    movl -4(%rbp), %eax
+    imul -72(%rbp), %eax
+    movl %eax, -76(%rbp)
+    movl -76(%rbp), %edi
     movq $0, %rax
     call get_int
-    movq %rax, -144(%rbp)
-    movq -120(%rbp), %rax
-    imul -144(%rbp), %rax
-    movq %rax, -152(%rbp)
-    movq -8(%rbp), %rax
-    imul -152(%rbp), %rax
-    movq %rax, -8(%rbp)
-    movq $21, %rbx
-    movq $0, %rax
-    shlq $32, %rax
-    orq %rax, %rbx
-    movq %rbx, -160(%rbp)
-    movq -24(%rbp), %rax
-    imul -160(%rbp), %rax
-    movq %rax, -168(%rbp)
-    movq -168(%rbp), %rdi
+    movl %eax, -80(%rbp)
+    movl $21, -88(%rbp)
+    movl -16(%rbp), %eax
+    imul -88(%rbp), %eax
+    movl %eax, -92(%rbp)
+    movl -92(%rbp), %edi
     movq $0, %rax
     call get_int
-    movq %rax, -176(%rbp)
-    movq $7, %rbx
-    movq $0, %rax
-    shlq $32, %rax
-    orq %rax, %rbx
-    movq %rbx, -184(%rbp)
-    movq -184(%rbp), %rdi
+    movl %eax, -96(%rbp)
+    movl $7, -100(%rbp)
+    movl -100(%rbp), %edi
     movq $0, %rax
     call get_int
-    movq %rax, -192(%rbp)
-    movq -176(%rbp), %rax
-    cqto
-    movq -192(%rbp), %rcx
-    idiv %rcx
-    movq %rax, -200(%rbp)
-    movq $15, %rbx
-    movq $0, %rax
-    shlq $32, %rax
-    orq %rax, %rbx
-    movq %rbx, -208(%rbp)
-    movq -208(%rbp), %rdi
+    movl %eax, -104(%rbp)
+    movl -96(%rbp), %eax
+    cdq
+    movl -104(%rbp), %ecx
+    idiv %ecx
+    movl %eax, -108(%rbp)
+    movl $15, -112(%rbp)
+    movl -112(%rbp), %edi
     movq $0, %rax
     call get_int
-    movq %rax, -216(%rbp)
-    movq -200(%rbp), %rax
-    addq -216(%rbp), %rax
-    movq %rax, -224(%rbp)
-    movq -8(%rbp), %rax
-    cqto
-    movq -224(%rbp), %rcx
-    idiv %rcx
-    movq %rax, -8(%rbp)
-    movq $1, %rbx
-    movq $0, %rax
-    shlq $32, %rax
-    orq %rax, %rbx
-    movq %rbx, -232(%rbp)
-    movq -8(%rbp), %rax
-    addq -232(%rbp), %rax
-    movq %rax, -240(%rbp)
-    movq $100, %rbx
-    movq $0, %rax
-    shlq $32, %rax
-    orq %rax, %rbx
-    movq %rbx, -248(%rbp)
-    movq -240(%rbp), %rax
-    imul -248(%rbp), %rax
-    movq %rax, -256(%rbp)
-    movq -256(%rbp), %rdi
+    movl %eax, -116(%rbp)
+    movl -108(%rbp), %eax
+    addl -116(%rbp), %eax
+    movl %eax, -120(%rbp)
+    movl -4(%rbp), %eax
+    cdq
+    movl -120(%rbp), %ecx
+    idiv %ecx
+    movl %eax, -4(%rbp)
+    movl $1, -124(%rbp)
+    movl -4(%rbp), %eax
+    addl -124(%rbp), %eax
+    movl %eax, -128(%rbp)
+    movl $100, -132(%rbp)
+    movl -128(%rbp), %eax
+    imul -132(%rbp), %eax
+    movl %eax, -136(%rbp)
+    movl -136(%rbp), %edi
     movq $0, %rax
     call get_int
-    movq %rax, -264(%rbp)
-    movq -8(%rbp), %rax
-    addq -264(%rbp), %rax
-    movq %rax, -8(%rbp)
-    movq $41, %rbx
-    movq $0, %rax
-    shlq $32, %rax
-    orq %rax, %rbx
-    movq %rbx, -272(%rbp)
-    movq -24(%rbp), %rax
-    imul -272(%rbp), %rax
-    movq %rax, -280(%rbp)
-    movq -280(%rbp), %rdi
+    movl %eax, -140(%rbp)
+    movl $41, -144(%rbp)
+    movl -16(%rbp), %eax
+    imul -144(%rbp), %eax
+    movl %eax, -148(%rbp)
+    movl -148(%rbp), %edi
     movq $0, %rax
     call get_int
-    movq %rax, -288(%rbp)
-    movq $18, %rbx
-    movq $0, %rax
-    shlq $32, %rax
-    orq %rax, %rbx
-    movq %rbx, -296(%rbp)
-    movq -296(%rbp), %rdi
+    movl %eax, -152(%rbp)
+    movl $18, -156(%rbp)
+    movl -156(%rbp), %edi
     movq $0, %rax
     call get_int
-    movq %rax, -304(%rbp)
-    movq -288(%rbp), %rax
-    cqto
-    movq -304(%rbp), %rcx
-    idiv %rcx
-    movq %rdx, -312(%rbp)
-    movq $15, %rbx
-    movq $0, %rax
-    shlq $32, %rax
-    orq %rax, %rbx
-    movq %rbx, -320(%rbp)
-    movq -320(%rbp), %rdi
+    movl %eax, -160(%rbp)
+    movl -152(%rbp), %eax
+    cdq
+    movl -160(%rbp), %ecx
+    idiv %ecx
+    movl %edx, -164(%rbp)
+    movl $15, -168(%rbp)
+    movl -168(%rbp), %edi
     movq $0, %rax
     call get_int
-    movq %rax, -328(%rbp)
-    movq -312(%rbp), %rax
-    addq -328(%rbp), %rax
-    movq %rax, -336(%rbp)
-    movq -8(%rbp), %rax
-    cqto
-    movq -336(%rbp), %rcx
-    idiv %rcx
-    movq %rdx, -8(%rbp)
-    movq -24(%rbp), %rax
-    movq %rax, -344(%rbp)
-    movq -344(%rbp), %rdi
+    movl %eax, -172(%rbp)
+    movl -164(%rbp), %eax
+    addl -172(%rbp), %eax
+    movl %eax, -176(%rbp)
+    movl -4(%rbp), %eax
+    cdq
+    movl -176(%rbp), %ecx
+    idiv %ecx
+    movl %edx, -4(%rbp)
+    movl -16(%rbp), %eax
+    movsxd %eax, %rax
+    movq %rax, -184(%rbp)
+    movq -184(%rbp), %rdi
     movq $0, %rax
     call get_long
-    movq %rax, -352(%rbp)
-    movq -16(%rbp), %rax
-    addq -352(%rbp), %rax
-    movq %rax, -16(%rbp)
-    movq -24(%rbp), %rax
-    movq %rax, -360(%rbp)
-    movq -360(%rbp), %rdi
+    movq %rax, -192(%rbp)
+    movl -16(%rbp), %eax
+    movsxd %eax, %rax
+    movq %rax, -200(%rbp)
+    movq -200(%rbp), %rdi
     movq $0, %rax
     call get_long
-    movq %rax, -368(%rbp)
-    movq -8(%rbp), %rax
-    movq %rax, -376(%rbp)
+    movq %rax, -208(%rbp)
+    movl -4(%rbp), %eax
+    movsxd %eax, %rax
+    movq %rax, -216(%rbp)
     movq $3, %rbx
     movq $0, %rax
     shlq $32, %rax
     orq %rax, %rbx
-    movq %rbx, -384(%rbp)
-    movq -376(%rbp), %rax
-    subq -384(%rbp), %rax
-    movq %rax, -392(%rbp)
-    movq -392(%rbp), %rdi
+    movq %rbx, -224(%rbp)
+    movq -216(%rbp), %rax
+    subq -224(%rbp), %rax
+    movq %rax, -232(%rbp)
+    movq -232(%rbp), %rdi
     movq $0, %rax
     call get_long
-    movq %rax, -400(%rbp)
-    movq -368(%rbp), %rax
-    subq -400(%rbp), %rax
-    movq %rax, -408(%rbp)
-    movq -16(%rbp), %rax
-    subq -408(%rbp), %rax
-    movq %rax, -16(%rbp)
-    movq -24(%rbp), %rax
-    movq %rax, -416(%rbp)
-    movq -416(%rbp), %rdi
+    movq %rax, -240(%rbp)
+    movl -16(%rbp), %eax
+    movsxd %eax, %rax
+    movq %rax, -256(%rbp)
+    movq -256(%rbp), %rdi
     movq $0, %rax
     call get_long
-    movq %rax, -424(%rbp)
-    movq -8(%rbp), %rax
-    movq %rax, -432(%rbp)
+    movq %rax, -264(%rbp)
+    movl -4(%rbp), %eax
+    movsxd %eax, %rax
+    movq %rax, -272(%rbp)
     movq $5, %rbx
     movq $0, %rax
     shlq $32, %rax
     orq %rax, %rbx
-    movq %rbx, -440(%rbp)
-    movq -432(%rbp), %rax
-    imul -440(%rbp), %rax
-    movq %rax, -448(%rbp)
-    movq -448(%rbp), %rdi
+    movq %rbx, -280(%rbp)
+    movq -272(%rbp), %rax
+    imul -280(%rbp), %rax
+    movq %rax, -288(%rbp)
+    movq -288(%rbp), %rdi
     movq $0, %rax
     call get_long
-    movq %rax, -456(%rbp)
-    movq -424(%rbp), %rax
-    imul -456(%rbp), %rax
-    movq %rax, -464(%rbp)
-    movq -16(%rbp), %rax
-    imul -464(%rbp), %rax
-    movq %rax, -16(%rbp)
-    movq -24(%rbp), %rax
-    movq %rax, -472(%rbp)
+    movq %rax, -296(%rbp)
+    movl -16(%rbp), %eax
+    movsxd %eax, %rax
+    movq %rax, -312(%rbp)
     movq $21, %rbx
     movq $0, %rax
     shlq $32, %rax
     orq %rax, %rbx
-    movq %rbx, -480(%rbp)
-    movq -472(%rbp), %rax
-    imul -480(%rbp), %rax
-    movq %rax, -488(%rbp)
-    movq -488(%rbp), %rdi
+    movq %rbx, -320(%rbp)
+    movq -312(%rbp), %rax
+    imul -320(%rbp), %rax
+    movq %rax, -328(%rbp)
+    movq -328(%rbp), %rdi
     movq $0, %rax
     call get_long
-    movq %rax, -496(%rbp)
+    movq %rax, -336(%rbp)
     movq $7, %rbx
     movq $0, %rax
     shlq $32, %rax
     orq %rax, %rbx
-    movq %rbx, -504(%rbp)
-    movq -504(%rbp), %rdi
+    movq %rbx, -344(%rbp)
+    movq -344(%rbp), %rdi
     movq $0, %rax
     call get_long
-    movq %rax, -512(%rbp)
-    movq -496(%rbp), %rax
-    cqto
-    movq -512(%rbp), %rcx
-    idiv %rcx
-    movq %rax, -520(%rbp)
+    movq %rax, -352(%rbp)
     movq $15, %rbx
     movq $0, %rax
     shlq $32, %rax
     orq %rax, %rbx
-    movq %rbx, -528(%rbp)
-    movq -528(%rbp), %rdi
+    movq %rbx, -368(%rbp)
+    movq -368(%rbp), %rdi
     movq $0, %rax
     call get_long
-    movq %rax, -536(%rbp)
-    movq -520(%rbp), %rax
-    addq -536(%rbp), %rax
-    movq %rax, -544(%rbp)
-    movq -16(%rbp), %rax
-    cqto
-    movq -544(%rbp), %rcx
-    idiv %rcx
-    movq %rax, -16(%rbp)
-    movq -8(%rbp), %rax
-    movq %rax, -552(%rbp)
+    movq %rax, -376(%rbp)
+    movl -4(%rbp), %eax
+    movsxd %eax, %rax
+    movq %rax, -392(%rbp)
     movq $1, %rbx
     movq $0, %rax
     shlq $32, %rax
     orq %rax, %rbx
-    movq %rbx, -560(%rbp)
-    movq -552(%rbp), %rax
-    addq -560(%rbp), %rax
-    movq %rax, -568(%rbp)
+    movq %rbx, -400(%rbp)
+    movq -392(%rbp), %rax
+    addq -400(%rbp), %rax
+    movq %rax, -408(%rbp)
     movq $100, %rbx
     movq $0, %rax
     shlq $32, %rax
     orq %rax, %rbx
-    movq %rbx, -576(%rbp)
-    movq -568(%rbp), %rax
-    imul -576(%rbp), %rax
-    movq %rax, -584(%rbp)
-    movq -584(%rbp), %rdi
+    movq %rbx, -416(%rbp)
+    movq -408(%rbp), %rax
+    imul -416(%rbp), %rax
+    movq %rax, -424(%rbp)
+    movq -424(%rbp), %rdi
     movq $0, %rax
     call get_long
-    movq %rax, -592(%rbp)
-    movq -16(%rbp), %rax
-    addq -592(%rbp), %rax
-    movq %rax, -16(%rbp)
-    movq -24(%rbp), %rax
-    movq %rax, -600(%rbp)
+    movq %rax, -432(%rbp)
+    movl -16(%rbp), %eax
+    movsxd %eax, %rax
+    movq %rax, -440(%rbp)
     movq $41, %rbx
     movq $0, %rax
     shlq $32, %rax
     orq %rax, %rbx
-    movq %rbx, -608(%rbp)
-    movq -600(%rbp), %rax
-    imul -608(%rbp), %rax
-    movq %rax, -616(%rbp)
-    movq -616(%rbp), %rdi
+    movq %rbx, -448(%rbp)
+    movq -440(%rbp), %rax
+    imul -448(%rbp), %rax
+    movq %rax, -456(%rbp)
+    movq -456(%rbp), %rdi
     movq $0, %rax
     call get_long
-    movq %rax, -624(%rbp)
+    movq %rax, -464(%rbp)
     movq $18, %rbx
     movq $0, %rax
     shlq $32, %rax
     orq %rax, %rbx
-    movq %rbx, -632(%rbp)
-    movq -632(%rbp), %rdi
+    movq %rbx, -472(%rbp)
+    movq -472(%rbp), %rdi
     movq $0, %rax
     call get_long
-    movq %rax, -640(%rbp)
-    movq -624(%rbp), %rax
+    movq %rax, -480(%rbp)
+    movq -464(%rbp), %rax
     cqto
-    movq -640(%rbp), %rcx
+    movq -480(%rbp), %rcx
     idiv %rcx
-    movq %rdx, -648(%rbp)
+    movq %rdx, -488(%rbp)
     movq $15, %rbx
     movq $0, %rax
     shlq $32, %rax
     orq %rax, %rbx
-    movq %rbx, -656(%rbp)
-    movq -656(%rbp), %rdi
+    movq %rbx, -496(%rbp)
+    movq -496(%rbp), %rdi
     movq $0, %rax
     call get_long
-    movq %rax, -664(%rbp)
-    movq -648(%rbp), %rax
-    addq -664(%rbp), %rax
-    movq %rax, -672(%rbp)
-    movq -16(%rbp), %rax
+    movq %rax, -504(%rbp)
+    movq -488(%rbp), %rax
+    addq -504(%rbp), %rax
+    movq %rax, -512(%rbp)
+    movq -12(%rbp), %rax
     cqto
-    movq -672(%rbp), %rcx
+    movq -512(%rbp), %rcx
     idiv %rcx
-    movq %rdx, -16(%rbp)
+    movq %rdx, -12(%rbp)
     jmp main3
 main3:
-    movq $3, %rbx
-    movq $0, %rax
-    shlq $32, %rax
-    orq %rax, %rbx
-    movq %rbx, -680(%rbp)
-    movq -24(%rbp), %rax
-    addq -680(%rbp), %rax
-    movq %rax, -24(%rbp)
+    movl $3, -516(%rbp)
+    movl -16(%rbp), %eax
+    addl -516(%rbp), %eax
+    movl %eax, -16(%rbp)
     jmp main1
 main4:
-    movq $0, %rbx
+    movl $0, -520(%rbp)
+    movl -520(%rbp), %eax
+    movq %rbp, %rsp
+    popq %rbp
+    ret
+
+get_long:
+    pushq %rbp
+    movq %rsp, %rbp
+    subq $16, %rsp
+get_long0:
+    movq %rdi, %rax
+    movq %rax, -8(%rbp)
+    leaq str1(%rip), %rax
+    movq %rax, -16(%rbp)
+    movq -16(%rbp), %rdi
+    movq -8(%rbp), %rsi
     movq $0, %rax
-    shlq $32, %rax
-    orq %rax, %rbx
-    movq %rbx, -688(%rbp)
-    movq -688(%rbp), %rax
+    call printf
+    movq -8(%rbp), %rax
     movq %rbp, %rsp
     popq %rbp
     ret
 
->>>>>>> dce
diff --git a/partition.dot b/partition.dot
index 5e77413..e98f379 100644
--- a/partition.dot
+++ b/partition.dot
@@ -8,8 +8,8 @@ digraph CFG {
   5 [label="Block 5\l_t8 <- 1\li <- i + _t8\l_t9 <- a[i]\l_t10 <- a[j]\la[i] <- _t10\la[j] <- _t9\lujmp partition6\l"];
   6 [label="Block 6\lujmp partition3\l"];
   0 -> 1 [label=""];
-  1 -> 4 [label="False"];
   1 -> 2 [label="True"];
+  1 -> 4 [label="False"];
   2 -> 6 [label="False"];
   2 -> 5 [label="True"];
   3 -> 1 [label=""];
diff --git a/phw.dot b/phw.dot
index 555be47..42838a2 100644
--- a/phw.dot
+++ b/phw.dot
@@ -1,50 +1,50 @@
 digraph CFG {
   node [shape=box fontname="monospace"];
   0 [label="Block 0\l_t0 <- arg0\l_t1 <- arg1\l_t2 <- arg2\l_t3 <- arg3\l_t4 <- arg4\l_t5 <- arg5\l_t6 <- arg6\l_t7 <- arg7\l_t8 <- arg8\l_t9 <- arg9\lcjmp _t0, phw1\lujmp phw2\l"];
-  1 [label="Block 1\l_t10 <- String(7, String)\lprintf(_t10)\lcjmp _t1, phw3\lujmp phw4\l"];
+  1 [label="Block 1\l_t10 <- String(0, String)\lprintf(_t10)\lcjmp _t1, phw3\lujmp phw4\l"];
   2 [label="Block 2\l_t23 <- not2(_t0)\lcjmp _t23, phw19\lujmp phw20\l"];
-  3 [label="Block 3\l_t11 <- String(8, String)\lprintf(_t11)\lcjmp _t2, phw5\lujmp phw6\l"];
+  3 [label="Block 3\l_t11 <- String(1, String)\lprintf(_t11)\lcjmp _t2, phw5\lujmp phw6\l"];
   4 [label="Block 4\l_t21 <- not2(_t1)\lcjmp _t21, phw17\lujmp phw18\l"];
-  5 [label="Block 5\l_t12 <- String(9, String)\lprintf(_t12)\lcjmp _t3, phw7\lujmp phw8\l"];
+  5 [label="Block 5\l_t12 <- String(2, String)\lprintf(_t12)\lcjmp _t3, phw7\lujmp phw8\l"];
   6 [label="Block 6\l_t19 <- not2(_t2)\lcjmp _t19, phw15\lujmp phw16\l"];
-  7 [label="Block 7\l_t13 <- String(10, String)\lprintf(_t13)\lcjmp _t4, phw9\lujmp phw10\l"];
+  7 [label="Block 7\l_t13 <- String(3, String)\lprintf(_t13)\lcjmp _t4, phw9\lujmp phw10\l"];
   8 [label="Block 8\l_t17 <- not2(_t3)\lcjmp _t17, phw13\lujmp phw14\l"];
-  9 [label="Block 9\l_t14 <- String(11, String)\lprintf(_t14)\lujmp phw10\l"];
+  9 [label="Block 9\l_t14 <- String(4, String)\lprintf(_t14)\lujmp phw10\l"];
   10 [label="Block 10\l_t15 <- not2(_t4)\lcjmp _t15, phw11\lujmp phw12\l"];
-  11 [label="Block 11\l_t16 <- String(12, String)\lprintf(_t16)\lujmp phw12\l"];
+  11 [label="Block 11\l_t16 <- String(5, String)\lprintf(_t16)\lujmp phw12\l"];
   12 [label="Block 12\lujmp phw8\l"];
-  13 [label="Block 13\l_t18 <- String(13, String)\lprintf(_t18)\lujmp phw14\l"];
+  13 [label="Block 13\l_t18 <- String(6, String)\lprintf(_t18)\lujmp phw14\l"];
   14 [label="Block 14\lujmp phw6\l"];
-  15 [label="Block 15\l_t20 <- String(14, String)\lprintf(_t20)\lujmp phw16\l"];
+  15 [label="Block 15\l_t20 <- String(7, String)\lprintf(_t20)\lujmp phw16\l"];
   16 [label="Block 16\lujmp phw4\l"];
-  17 [label="Block 17\l_t22 <- String(15, String)\lprintf(_t22)\lujmp phw18\l"];
+  17 [label="Block 17\l_t22 <- String(8, String)\lprintf(_t22)\lujmp phw18\l"];
   18 [label="Block 18\lujmp phw2\l"];
-  19 [label="Block 19\l_t24 <- String(16, String)\lprintf(_t24)\lujmp phw20\l"];
-  20 [label="Block 20\l_t25 <- String(17, String)\lprintf(_t25)\l_t26 <- not2(_t5)\lcjmp _t26, phw21\lujmp phw22\l"];
-  21 [label="Block 21\l_t27 <- String(18, String)\lprintf(_t27)\lujmp phw22\l"];
+  19 [label="Block 19\l_t24 <- String(9, String)\lprintf(_t24)\lujmp phw20\l"];
+  20 [label="Block 20\l_t25 <- String(10, String)\lprintf(_t25)\l_t26 <- not2(_t5)\lcjmp _t26, phw21\lujmp phw22\l"];
+  21 [label="Block 21\l_t27 <- String(11, String)\lprintf(_t27)\lujmp phw22\l"];
   22 [label="Block 22\lcjmp _t6, phw23\lujmp phw24\l"];
   23 [label="Block 23\l_t28 <- not2(_t7)\lcjmp _t28, phw25\lujmp phw26\l"];
   24 [label="Block 24\l_t32 <- not2(_t6)\lcjmp _t32, phw31\lujmp phw32\l"];
-  25 [label="Block 25\l_t29 <- String(19, String)\lprintf(_t29)\lcjmp _t8, phw27\lujmp phw28\l"];
+  25 [label="Block 25\l_t29 <- String(12, String)\lprintf(_t29)\lcjmp _t8, phw27\lujmp phw28\l"];
   26 [label="Block 26\lujmp phw24\l"];
-  27 [label="Block 27\l_t30 <- String(20, String)\lprintf(_t30)\lcjmp _t9, phw29\lujmp phw30\l"];
+  27 [label="Block 27\l_t30 <- String(13, String)\lprintf(_t30)\lcjmp _t9, phw29\lujmp phw30\l"];
   28 [label="Block 28\lujmp phw26\l"];
-  29 [label="Block 29\l_t31 <- String(21, String)\lprintf(_t31)\lujmp phw30\l"];
+  29 [label="Block 29\l_t31 <- String(14, String)\lprintf(_t31)\lujmp phw30\l"];
   30 [label="Block 30\lujmp phw28\l"];
-  31 [label="Block 31\l_t33 <- String(22, String)\lprintf(_t33)\lujmp phw32\l"];
-  32 [label="Block 32\l_t34 <- String(23, String)\lprintf(_t34)\l_t35 <- String(24, String)\lprintf(_t35)\l"];
+  31 [label="Block 31\l_t33 <- String(15, String)\lprintf(_t33)\lujmp phw32\l"];
+  32 [label="Block 32\l_t34 <- String(16, String)\lprintf(_t34)\l_t35 <- String(17, String)\lprintf(_t35)\l"];
   0 -> 1 [label="True"];
   0 -> 2 [label="False"];
-  1 -> 4 [label="False"];
   1 -> 3 [label="True"];
+  1 -> 4 [label="False"];
   2 -> 19 [label="True"];
   2 -> 20 [label="False"];
   3 -> 5 [label="True"];
   3 -> 6 [label="False"];
-  4 -> 17 [label="True"];
   4 -> 18 [label="False"];
-  5 -> 8 [label="False"];
+  4 -> 17 [label="True"];
   5 -> 7 [label="True"];
+  5 -> 8 [label="False"];
   6 -> 15 [label="True"];
   6 -> 16 [label="False"];
   7 -> 10 [label="False"];
diff --git a/phw.png b/phw.png
index 9935c03..91c6c63 100644
Binary files a/phw.png and b/phw.png differ
diff --git a/recursive.dot b/recursive.dot
index ade70a0..07ad321 100644
--- a/recursive.dot
+++ b/recursive.dot
@@ -6,14 +6,14 @@ digraph CFG {
   3 [label="Block 3\l_t9 <- 1\lx <- x + _t9\lujmp recursive1\l"];
   4 [label="Block 4\l_t10 <- 0\l_t11 <- _t0 == _t10\lcjmp _t11, recursive5\lujmp recursive6\l"];
   5 [label="Block 5\l_t12 <- 0\lret _t12\l"];
-  6 [label="Block 6\l_t13 <- String(2, String)\l_t16 <- 15\l_t17 <- _t0 % _t16\l_t15 <- Int(_t17)\l_t14 <- i[_t15]\lprintf(_t13, _t14)\l_t18 <- 1\l_t19 <- _t0 - _t18\l_t20 <- recursive(_t19)\l_t23 <- 15\l_t24 <- _t0 % _t23\l_t22 <- Int(_t24)\l_t21 <- i[_t22]\l_t25 <- _t20 + _t21\lret _t25\l"];
+  6 [label="Block 6\l_t13 <- String(0, String)\l_t16 <- 15\l_t17 <- _t0 % _t16\l_t15 <- Int(_t17)\l_t14 <- i[_t15]\lprintf(_t13, _t14)\l_t18 <- 1\l_t19 <- _t0 - _t18\l_t20 <- recursive(_t19)\l_t23 <- 15\l_t24 <- _t0 % _t23\l_t22 <- Int(_t24)\l_t21 <- i[_t22]\l_t25 <- _t20 + _t21\lret _t25\l"];
   7 [label="Block 7\lexit(-1)\l"];
   -1 -> 7 [label=""];
   0 -> 1 [label=""];
-  1 -> 2 [label="True"];
   1 -> 4 [label="False"];
+  1 -> 2 [label="True"];
   2 -> 3 [label=""];
   3 -> 1 [label=""];
-  4 -> 5 [label="True"];
   4 -> 6 [label="False"];
+  4 -> 5 [label="True"];
 }
diff --git a/recursive.png b/recursive.png
index 4a7c792..710ccb8 100644
Binary files a/recursive.png and b/recursive.png differ
diff --git a/second.dot b/second.dot
index 0289a14..40efcec 100644
--- a/second.dot
+++ b/second.dot
@@ -1,4 +1,4 @@
 digraph CFG {
   node [shape=box fontname="monospace"];
-  0 [label="Block 0\l_t0 <- String(0, String)\lprintf(_t0)\l_t1 <- 2\lret _t1\l"];
+  0 [label="Block 0\l_t0 <- String(2, String)\lprintf(_t0)\l_t1 <- 2\lret _t1\l"];
 }
diff --git a/second.png b/second.png
index b1e0a7c..dcf7dcc 100644
Binary files a/second.png and b/second.png differ
diff --git a/src/dataflow.rs b/src/dataflow.rs
index 001a732..8b136d4 100644
--- a/src/dataflow.rs
+++ b/src/dataflow.rs
@@ -254,6 +254,8 @@ fn dead_code_elimination(method_cfg: &mut CFG, debug: bool) -> bool {
     let (_, out_maps) = compute_liveness(method_cfg, debug);
     let mut update_occurred = false;
 
+    println!("liveness is {:#?}", out_maps);
+
 
     for (block_id, block) in method_cfg.blocks.iter_mut() {
         let out_set = out_maps.get(&block_id).unwrap_or(&HashSet::new()).clone();
@@ -278,7 +280,9 @@ fn dead_code_elimination(method_cfg: &mut CFG, debug: bool) -> bool {
                 // Any write to global var is a side effect
                  _=> {
                     if let Some(dest) = get_dest_var(instr) {
-                        method_cfg.locals.get(dest.as_str()).is_none()
+                        let is_global_write = method_cfg.locals.get(dest.as_str()).is_none();
+                        // println!("{dest} is {} a global write", is_global_write);
+                        is_global_write
                     } else {
                         false
                     }
@@ -294,14 +298,18 @@ fn dead_code_elimination(method_cfg: &mut CFG, debug: bool) -> bool {
             if keep_instr {
                 // Update live set
                 for var in &used {
-                    live.insert(var.clone());
+                    // Not live if assigning to self
+                    if Some(var) != dest.as_ref() {
+                        live.insert(var.clone());
+                    }
+
                 }
 
                 if let Some(var) = &dest {
                     // TODO: Don't (?) remove from liveness if is a def and a use
-                    if !&used.contains(var) {
+                    // if !&used.contains(var) {
                         live.remove(var);
-                    }
+                    // }
                 }
                 new_instrs.push(instr.clone());
             } else {
diff --git a/sub.dot b/sub.dot
index 0c001b5..942cbf9 100644
--- a/sub.dot
+++ b/sub.dot
@@ -1,4 +1,4 @@
 digraph CFG {
   node [shape=box fontname="monospace"];
-  0 [label="Block 0\l_t0 <- arg0\l_t1 <- arg1\l_t2 <- _t1 - _t0\lret _t2\l"];
+  0 [label="Block 0\l_t0 <- arg0\l_t1 <- arg1\l_t2 <- _t0 - _t1\lret _t2\l"];
 }
diff --git a/sub.png b/sub.png
index 1627781..98d299e 100644
Binary files a/sub.png and b/sub.png differ
diff --git a/test_long_args.dot b/test_long_args.dot
index ec9f042..b8a0589 100644
--- a/test_long_args.dot
+++ b/test_long_args.dot
@@ -1,4 +1,4 @@
 digraph CFG {
   node [shape=box fontname="monospace"];
-  0 [label="Block 0\l_t0 <- arg0\l_t1 <- arg1\l_t2 <- arg2\l_t3 <- arg3\l_t4 <- arg4\l_t5 <- arg5\l_t6 <- String(2, String)\lprintf(_t6, _t0, _t1, _t2, _t3, _t4)\l_t7 <- String(3, String)\l_t8 <- _t0 + _t1\l_t9 <- _t8 + _t2\l_t10 <- _t9 + _t3\l_t11 <- _t10 + _t4\l_t12 <- _t11 + _t5\lprintf(_t7, _t5, _t12)\l_t17 <- _t12\lret _t17\l"];
+  0 [label="Block 0\l_t0 <- arg0\l_t1 <- arg1\l_t2 <- arg2\l_t3 <- arg3\l_t4 <- arg4\l_t5 <- arg5\l_t6 <- String(0, String)\lprintf(_t6, _t0, _t1, _t2, _t3, _t4)\l_t7 <- String(1, String)\l_t8 <- _t0 + _t1\l_t9 <- _t8 + _t2\l_t10 <- _t9 + _t3\l_t11 <- _t10 + _t4\l_t12 <- _t11 + _t5\lprintf(_t7, _t5, _t12)\l_t17 <- _t12\lret _t17\l"];
 }
diff --git a/test_long_args.png b/test_long_args.png
index 2ccd016..79f13ac 100644
Binary files a/test_long_args.png and b/test_long_args.png differ
diff --git a/test_short_circuit.dot b/test_short_circuit.dot
index 82fabf1..aec53eb 100644
--- a/test_short_circuit.dot
+++ b/test_short_circuit.dot
@@ -3,11 +3,11 @@ digraph CFG {
   0 [label="Block 0\l_t0 <- arg0\l_t1 <- arg1\l_t2 <- _t1 != 1\lcjmp _t2, test_short_circuit1\lujmp test_short_circuit2\l"];
   1 [label="Block 1\l_t3 <- _t0 != 1\lcjmp _t3, test_short_circuit3\lujmp test_short_circuit4\l"];
   2 [label="Block 2\lret 1\l"];
-  3 [label="Block 3\l_t4 <- String(0, String)\lprintf(_t4)\lujmp test_short_circuit5\l"];
-  4 [label="Block 4\l_t5 <- String(1, String)\lprintf(_t5)\lujmp test_short_circuit5\l"];
+  3 [label="Block 3\l_t4 <- String(2, String)\lprintf(_t4)\lujmp test_short_circuit5\l"];
+  4 [label="Block 4\l_t5 <- String(3, String)\lprintf(_t5)\lujmp test_short_circuit5\l"];
   5 [label="Block 5\lujmp test_short_circuit2\l"];
-  0 -> 1 [label="True"];
   0 -> 2 [label="False"];
+  0 -> 1 [label="True"];
   1 -> 3 [label="True"];
   1 -> 4 [label="False"];
   3 -> 5 [label=""];
diff --git a/test_short_circuit.png b/test_short_circuit.png
index 1319801..86ab6fa 100644
Binary files a/test_short_circuit.png and b/test_short_circuit.png differ
diff --git a/tests/phase3/input/sp25-10-side-effects.dcf b/tests/phase3/input/sp25-10-side-effects.dcf
index e3248de..dc6c87f 100644
--- a/tests/phase3/input/sp25-10-side-effects.dcf
+++ b/tests/phase3/input/sp25-10-side-effects.dcf
@@ -22,17 +22,11 @@ void main () {
         a -= get_int(i) - get_int(a - 3);
         a *= get_int(i) * get_int(a * 5);
 
-
         a /= (get_int(i * 21) / get_int(7)) + get_int(15);
 
-
-
-
         a += get_int((a+1) * 100);
         a %= (get_int(i * 41) % get_int(18)) + get_int(15);
 
-
-
         b += get_long(long(i));
         b -= get_long(long(i)) - get_long(long(a) - 3L);
         b *= get_long(long(i)) * get_long(long(a) * 5L);
diff --git a/third.dot b/third.dot
index 2345c0a..5484a28 100644
--- a/third.dot
+++ b/third.dot
@@ -1,4 +1,4 @@
 digraph CFG {
   node [shape=box fontname="monospace"];
-  0 [label="Block 0\l_t0 <- String(2, String)\lprintf(_t0)\l_t1 <- 3\lret _t1\l"];
+  0 [label="Block 0\l_t0 <- String(4, String)\lprintf(_t0)\l_t1 <- 3\lret _t1\l"];
 }
diff --git a/third.png b/third.png
index beee681..04050ad 100644
Binary files a/third.png and b/third.png differ
